spec/std/humanize_spec.cr

```diff
@@ -29,6 +38,26 @@ describe Number do
       123_456.format(delimiter: '.').should eq "123.456"
 
       123_456.789.format.should eq "123,456.789"
+
+      1e15.format(decimal_places: 7).should eq "1,000,000,000,000,000.0000000"
+      1e15.to_i64.format(decimal_places: 7).should eq "1,000,000,000,000,000.0000000"
+      1e-5.format(decimal_places: 7).should eq "0.0000100"
+      1e-4.format(decimal_places: 7).should eq "0.0001000"
+
+      Float64::MAX.format.should eq "179,769,313,486,231,570,814,527,423,731,704,356,798,070,567,525,844,996,598,917,476,803,157,260,780,028,538,760,589,558,632,766,878,171,540,458,953,514,382,464,234,321,326,889,464,182,768,467,546,703,537,516,986,049,910,576,551,282,076,245,490,090,389,328,944,075,868,508,455,133,942,304,583,236,903,222,948,165,808,559,332,123,348,274,797,826,204,144,723,168,738,177,180,919,299,881,250,404,026,184,124,858,368.0"
+      Float64::MIN.format.should eq "-179,769,313,486,231,570,814,527,423,731,704,356,798,070,567,525,844,996,598,917,476,803,157,260,780,028,538,760,589,558,632,766,878,171,540,458,953,514,382,464,234,321,326,889,464,182,768,467,546,703,537,516,986,049,910,576,551,282,076,245,490,090,389,328,944,075,868,508,455,133,942,304,583,236,903,222,948,165,808,559,332,123,348,274,797,826,204,144,723,168,738,177,180,919,299,881,250,404,026,184,124,858,368.0"
+      Float64::MAX.format(decimal_places: 0).should eq "179,769,313,486,231,570,814,527,423,731,704,356,798,070,567,525,844,996,598,917,476,803,157,260,780,028,538,760,589,558,632,766,878,171,540,458,953,514,382,464,234,321,326,889,464,182,768,467,546,703,537,516,986,049,910,576,551,282,076,245,490,090,389,328,944,075,868,508,455,133,942,304,583,236,903,222,948,165,808,559,332,123,348,274,797,826,204,144,723,168,738,177,180,919,299,881,250,404,026,184,124,858,368"
+      Float64::MIN.format(decimal_places: 0).should eq "-179,769,313,486,231,570,814,527,423,731,704,356,798,070,567,525,844,996,598,917,476,803,157,260,780,028,538,760,589,558,632,766,878,171,540,458,953,514,382,464,234,321,326,889,464,182,768,467,546,703,537,516,986,049,910,576,551,282,076,245,490,090,389,328,944,075,868,508,455,133,942,304,583,236,903,222,948,165,808,559,332,123,348,274,797,826,204,144,723,168,738,177,180,919,299,881,250,404,026,184,124,858,368"
+      Float64::MIN_POSITIVE.format.should eq "0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000022250738585072014"
```

Comment:
I would expect this to be

```
0.000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000022250738585
07201383090232717332404064219215980462331830553327416887204434813918195854283159
01251102056406733973103581100515243416155346010885601238537771882113077799353200
23304796101474425836360719215650469425037342083752508066506166581589487204911799
68591639648500635908770118304874799780887753749949451580451605050915399856582470
81864511353793580499211598108576605199243335211435239014879569960959128889160299
26415110634663133936634775865130293717620473256317814856643508721228286376420448
46811407613911477062801689853244110024161447421618567166150540154285084716752901
90316132277889672970737312333408698898317506783884692609277397797285865965494109
1369095406136467568702398678315290680984617210924625396728515625
```

which would be 307 `0`s after the decimal point, followed by 715 significant digits.

If we proceed with this, an upper bound on the number of significant digits (of this exact representation) in a `Float::Primitive` can be found as follows:

```crystal
# this is slightly suboptimal for denormal values, but that probably
# doesn't matter in this context because those strings must be very long anyway
significand, exponent = Math.frexp(self)
significand = Math.ldexp(significand, Float64::MANT_DIGITS).to_i64
exponent -= Float64::MANT_DIGITS

# for positive exponent:
#   x == significand * 2.0 ** exponent
# for negative exponent:
#   x == significand * 5.0 ** -exponent * 10.0 ** exponent
sig_digits = Math.log10(significand) + Math.log10(2) * exponent
sig_digits -= exponent if exponent < 0
sig_digits = (sig_digits + 1).floor.to_i # or + 2, to avoid rounding errors
```

Then we also use `Math.log10(self)` to determine the number of leading zeros in the fractional part, or the number of significant digits that end up in the integral part. This gives us a big enough decimal place for `%f` and finally the trailing zeros are trimmed.

Reply:
That algorithm gives too much precision for values that are supposed to round to simple decimal numbers:

```cr
123.45.format # => "123.4500000000000028421709430404007434844970703125"
```

Reply:
Which values, then, are supposed not to give that much precision?

Reply:
I don't know. But I'm pretty sure that `123.45.format` should be expected to be `123.45`.

Reply:
On those grounds would `1.2345e+21.format` be `1234500000000000000000.0` or `1234499999999999868928.0`?

Reply:
Good point. Without putting much thought into it, I'd probably pick the latter. It's just human perception that the integer part is fully exact. Decimal places are expected to be rounded to imitate lower precision.

I can see this legitimately go in different directions. I personally don't really care much about the specific detailed decisions. Floating point values really isn't much of my home turf.

But I think it's paramount to fix the clearly incorrect behaviour of #10841 and `1.format(decimal_places: 1, only_significant: true) # => "1."` (also included in this PR). I'd prefer to merge as is, even with the current trivial and inefficient implementation, just to make things at least work in a reliable way.
Afterwards we can discuss the more detailed specifics of floating point formats.